### PR TITLE
BMC/IPMI smart proxy actions

### PR DESCRIPTION
**Pull request by dLobatog**
This pull request contains everything that is needed to perform IPMI operations on a BMC interface. In order to test this, a smart proxy with the BMC capability should be hooked to foreman. This should facilitate the integration of a console for physical machines in foreman.

A further step down the road would be to merge this and similar actions for virtual machines into something generic. 

After this is merged a host with IPMI enabled (username and password set) and a IPMI SmartProxy available, should look like this https://docs.google.com/file/d/0Bz4WMhz8WLjnaFpzdGFEMEQzY00/edit?usp=sharing

### CODE REVIEW

**Comment by GregSutcliffe** [comment]
[test]

**Comment by domcleal** [comment]
@GregSutcliffe could you review please (at least from a BMC, rather than UI perspective)?

@eLobato this looks nice, thanks!

**Comment by GregSutcliffe** [comment]
I've already had a brief look through, hence why I've marked it for testing. I'll give it closer scrutiny later on (possibly tomorrow morning).

File: app/helpers/bmc_helper.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
+module BmcHelper
+  def bmc_available?  
+    ipmi = @host.interfaces.select { |i| i.attrs[:provider] == "IPMI" }.first
```
Comment (ohadlevy) [inline app/helpers/bmc_helper.rb:3]: you could add here @host.interfaces.bmc.select .. to restirct only to bmc based interfaces

File: app/helpers/bmc_helper.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    if ipmi.present? 
+      true if ipmi.password.present? && ipmi.username.present? && ipmi_available?
```
Comment (ohadlevy) [inline app/helpers/bmc_helper.rb:5]: if the if statement return true then there is no point in explicitly saying true ;)

File: app/helpers/bmc_helper.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    if ipmi.present? 
+      true if ipmi.password.present? && ipmi.username.present? && ipmi_available? 
+    else
+      false
+    end
+  end
+
+  def ipmi_available?
+    begin 
+      timeout(15) do
```
Comment (ohadlevy) [inline app/helpers/bmc_helper.rb:13]: should the timeout be in the helper or maybe in the underling lib?

File: app/models/host/managed.rb [hunk 4]
```diff
@@ -731,6 +731,23 @@ def import_missing_ids
     tax_organization.import_missing_ids if organization
   end
 
+  def get_bmc_interface
+    url       = SmartProxy.select { |f| f.features.map(&:name).include? "BMC" }.first.url
```
Comment (ohadlevy) [inline app/models/host/managed.rb:735]: I guess this is here where it gets tricky, I would try to find one based on the interface subnet first?

**Comment by ohadlevy** [comment]
@eLobato awesome! thanks, 2 things worth thinking of:
1. how much effort would it be to add a simple wrapper class with start stop status etc ? this way we could elliminate the extra routes and simply relay on the existing power methods.
2. maybe we should extend the set build option, to have a checkbox to reboot / poweron a machine that supports it?

File: app/helpers/bmc_helper.rb (hunk 3 above)
Comment (dLobatog) [inline app/helpers/bmc_helper.rb:13]: in order for the timeout to be in the lib, basically all calls using "resource" in lib/proxy_api/resource.rb should include a timeout param, honestly I have no preference, the only advantage I find to put it on the helper is that you only explicitly control the timeout once (to make sure there's a connection), then the rest of the calls issued after that assume the proxy is available.

**Comment by dLobatog** [comment]
1. I'm not sure, what would this entail, checking for the host type (vm/physical) on the controller, then running the proper call to ipmi or compute_resource_vm? That'd get rid of the routes, but maybe you meant something different.
2. I'll look into it but maybe that can be a separate pull request

**Comment by ohadlevy** [comment]
@eLobato OK, also note that when adding new controller actions, you need to update the permissions as well (hence the failed tests)

**Comment by ohadlevy** [comment]
while testing this patch, i had to setup my bmc infra, and it seems like it does not handle errors gracefully.
maybe we should move to ajax instead? might make sense for compute resources too

**Comment by dLobatog** [comment]
What kind of errors did you get?

**Comment by ohadlevy** [comment]
for example, timeout would fail the entire action.
also, failures in dhcp, e.g. 400 Bad Request would fail the entire action.

**Comment by ohadlevy** [comment]
i've made some modifications, what do you think of https://github.com/ohadlevy/foreman/compare/556

**Comment by dLobatog** [comment]
I'll test this on friday, I'm on leave right now.

**Comment by ohadlevy** [comment]
closed in favour of #596

**Pull request closed without merging**